core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala

```diff
@@ -128,41 +130,52 @@ private[spark] class CompressedMapStatus(
  * @param numNonEmptyBlocks the number of non-empty blocks
  * @param emptyBlocks a bitmap tracking which blocks are empty
  * @param avgSize average size of the non-empty blocks
+ * @param hugeBlockSizesArray sizes of huge blocks by their reduceId.
  */
 private[spark] class HighlyCompressedMapStatus private (
     private[this] var loc: BlockManagerId,
     private[this] var numNonEmptyBlocks: Int,
     private[this] var emptyBlocks: RoaringBitmap,
-    private[this] var avgSize: Long)
+    private[this] var avgSize: Long,
+    private[this] var hugeBlockSizesArray: Array[Tuple2[Int, Byte]])
   extends MapStatus with Externalizable {
 
+  @transient var hugeBlockSizes: Map[Int, Byte] =
+    if (hugeBlockSizesArray == null) null else hugeBlockSizesArray.toMap
+
   // loc could be null when the default constructor is called during deserialization
   require(loc == null || avgSize > 0 || numNonEmptyBlocks == 0,
     "Average size can only be zero for map stages that produced no output")
 
-  protected def this() = this(null, -1, null, -1)  // For deserialization only
+  def this() = this(null, -1, null, -1, null)  // For deserialization only
 
   override def location: BlockManagerId = loc
 
   override def getSizeForBlock(reduceId: Int): Long = {
     if (emptyBlocks.contains(reduceId)) {
       0
     } else {
-      avgSize
+      hugeBlockSizes.get(reduceId) match {
+        case Some(size) => MapStatus.decompressSize(size)
+        case None => avgSize
+      }
     }
   }
 
   override def writeExternal(out: ObjectOutput): Unit = Utils.tryOrIOException {
     loc.writeExternal(out)
     emptyBlocks.writeExternal(out)
     out.writeLong(avgSize)
+    out.writeObject(hugeBlockSizesArray)
   }
 
   override def readExternal(in: ObjectInput): Unit = Utils.tryOrIOException {
     loc = BlockManagerId(in)
     emptyBlocks = new RoaringBitmap()
     emptyBlocks.readExternal(in)
     avgSize = in.readLong()
+    hugeBlockSizesArray = in.readObject().asInstanceOf[Array[Tuple2[Int, Byte]]]
+    hugeBlockSizes = hugeBlockSizesArray.toMap
```

Comment:
After the constructor is called, both `hugeBlockSizes` and `hugeBlockSizesArray` are initialized to be null, I think we need to initialize both in `readExternal` when deserialize.